File: Data Structure/20255/Solutions/Delete_Largest_Element_from_Max_Heap.py
class MaxHeap:
    def __init__(self):
        self.heap = []
    
    def insert(self, val):
        self.heap.append(val)
        self._heapify_up(len(self.heap) - 1)
    
    def _heapify_up(self, idx):
        parent = (idx - 1) // 2
        if idx > 0 and self.heap[idx] > self.heap[parent]:
            self.heap[idx], self.heap[parent] = self.heap[parent], self.heap[idx]
            self._heapify_up(parent)
    
    def deleteMax(self):
        if not self.heap:
            return None
        max_val = self.heap[0]
        self.heap[0] = self.heap[-1]
        self.heap.pop()
        idx = 0
        while idx < len(self.heap):
            largest = idx
            left = 2 * idx + 1
            right = 2 * idx + 2
            
            if left < len(self.heap) and self.heap[left] > self.heap[largest]:
                largest = left
            if right < len(self.heap) and self.heap[right] > self.heap[largest]:
                largest = right
            
            if largest == idx:
                break
            self.heap[idx], self.heap[largest] = self.heap[largest], self.heap[idx]
            idx = largest
        return max_val
    
    def getHeap(self):
        return self.heap

def solve():
    n = int(input())
    arr = list(map(int, input().split()))
    heap = MaxHeap()
    for num in arr:
        heap.insert(num)
    deleted = heap.deleteMax()
    remaining = heap.getHeap()
    print(deleted)
    if not remaining:
        print("Heap is Empty")
    else:
        print(" ".join(map(str, remaining)))

if __name__ == '__main__':
    solve()

File: Data Structure/20255/Solutions/Delete_Largest_Element_from_Max_Heap.js
class MaxHeap {
    constructor() {
        this.heap = [];
    }
    
    insert(val) {
        this.heap.push(val);
        this.heapifyUp(this.heap.length - 1);
    }
    
    heapifyUp(idx) {
        if (idx === 0) return;
        const parent = Math.floor((idx - 1) / 2);
        if (this.heap[idx] > this.heap[parent]) {
            [this.heap[idx], this.heap[parent]] = [this.heap[parent], this.heap[idx]];
            this.heapifyUp(parent);
        }
    }
    
    deleteMax() {
        if (this.heap.length === 0) return null;
        const maxVal = this.heap[0];
        this.heap[0] = this.heap[this.heap.length - 1];
        this.heap.pop();
        let idx = 0;
        while (idx < this.heap.length) {
            let largest = idx;
            const left = 2 * idx + 1;
            const right = 2 * idx + 2;
            
            if (left < this.heap.length && this.heap[left] > this.heap[largest]) {
                largest = left;
            }
            if (right < this.heap.length && this.heap[right] > this.heap[largest]) {
                largest = right;
            }
            
            if (largest === idx) break;
            [this.heap[idx], this.heap[largest]] = [this.heap[largest], this.heap[idx]];
            idx = largest;
        }
        return maxVal;
    }
    
    getHeap() {
        return this.heap;
    }
}

function solve(input) {
    const lines = input.trim().split("\n");
    const n = parseInt(lines[0]);
    const arr = lines[1].split(" ").map(Number);
    const heap = new MaxHeap();
    for (let num of arr) heap.insert(num);
    const deleted = heap.deleteMax();
    const remaining = heap.getHeap();
    console.log(deleted);
    if (remaining.length === 0) console.log("Heap is Empty");
    else console.log(remaining.join(" "));
}

const fs = require('fs');
solve(fs.readFileSync(0, 'utf-8'));

